File: Sources/KnowledgeMaster/AIClient.swift
import Foundation

enum AIClient {
    struct Request: Encodable {
        var model: String
        var messages: [Message]
        var temperature: Double = 0.2
    }

    struct VisionRequest: Encodable {
        var model: String
        var messages: [VisionMessage]
        var temperature: Double = 0.2
    }

    struct VisionMessage: Encodable {
        struct Part: Encodable {
            struct ImageURL: Encodable { var url: String }
            var type: String
            var text: String?
            var imageURL: ImageURL?

            enum CodingKeys: String, CodingKey {
                case type, text
                case imageURL = "image_url"
            }
        }

        var role: String
        var content: Content

        enum Content: Encodable {
            case text(String)
            case image(text: String, data: Data)

            func encode(to encoder: Encoder) throws {
                var container = encoder.singleValueContainer()
                switch self {
                case .text(let value):
                    try container.encode(value)
                case .image(let text, let data):
                    try container.encode([
                        Part(type: "text", text: text, imageURL: nil),
                        Part(type: "image_url", text: nil,
                             imageURL: .init(url: "data:image/png;base64,\(data.base64EncodedString())"))
                    ])
                }
            }
        }
    }

    struct Message: Codable, Hashable {
        var role: String
        var content: String
    }

    struct Response: Decodable {
        struct Choice: Decodable {
            var message: Message
        }
        var choices: [Choice]
    }

    struct ToolCall: Codable, Hashable {
        struct Function: Codable, Hashable {
            var name: String
            var arguments: String
        }

        var id: String
        var type: String
        var function: Function
    }

    struct ReActMessage: Codable, Hashable {
        var role: String
        var content: String?
        var reasoningContent: String?
        var toolCalls: [ToolCall]?
        var toolCallID: String?

        enum CodingKeys: String, CodingKey {
            case role, content
            case reasoningContent = "reasoning_content"
            case toolCalls = "tool_calls"
            case toolCallID = "tool_call_id"
        }

        init(role: String, content: String? = nil, reasoningContent: String? = nil,
             toolCalls: [ToolCall]? = nil, toolCallID: String? = nil) {
            self.role = role
            self.content = content
            self.reasoningContent = reasoningContent
            self.toolCalls = toolCalls
            self.toolCallID = toolCallID
        }
    }

    struct ToolDefinition: Encodable, Hashable {
        struct Function: Encodable, Hashable {
            var name: String
            var description: String
            var parameters: Parameters
        }

        struct Parameters: Encodable, Hashable {
            var type = "object"
            var properties: [String: Property]
            var required: [String]?
            var additionalProperties = false

            enum CodingKeys: String, CodingKey {
                case type, properties, required
                case additionalProperties = "additionalProperties"
            }
        }

        struct Property: Encodable, Hashable {
            var type: String
            var description: String
            var minimum: Int?
            var maximum: Int?
        }

        var type = "function"
        var function: Function
    }

    struct ReActRequest: Encodable {
        var model: String
        var messages: [ReActMessage]
        var tools: [ToolDefinition]
        var toolChoice = "auto"
        var temperature: Double = 0.2

        enum CodingKeys: String, CodingKey {
            case model, messages, tools, temperature
            case toolChoice = "tool_choice"
        }
    }

    struct ReActVisionRequest: Encodable {
        var model: String
        var messages: [ReActVisionMessage]
        var tools: [ToolDefinition]
        var toolChoice = "auto"
        var temperature: Double = 0.2

        enum CodingKeys: String, CodingKey {
            case model, messages, tools, temperature
            case toolChoice = "tool_choice"
        }
    }

    struct ReActVisionMessage: Encodable {
        var role: String
        var content: VisionMessage.Content?
        var reasoningContent: String?
        var toolCalls: [ToolCall]?
        var toolCallID: String?

        enum CodingKeys: String, CodingKey {
            case role, content
            case reasoningContent = "reasoning_content"
            case toolCalls = "tool_calls"
            case toolCallID = "tool_call_id"
        }
    }

    struct ReActResponse: Decodable {
        struct Choice: Decodable { var message: ReActMessage }
        var choices: [Choice]
    }

    struct ReActResult: Hashable {
        var answer: String
        var generatedFiles: [String]
    }

    static let toolDefinitions: [ToolDefinition] = [
        ToolDefinition(function: .init(
            name: "list_files",
            description: "列出已授权的知识库文件。documents/ 是只读资料，generated/ 是生成文件目录。",
            parameters: .init(properties: [
                "path": .init(type: "string", description: "可选的相对目录，如 documents、generated 或 .", minimum: nil, maximum: nil)
            ], required: nil)
        )),
        ToolDefinition(function: .init(
            name: "read_file",
            description: "按字符区间读取一个已授权文件。长文档应分段读取。",
            parameters: .init(properties: [
                "path": .init(type: "string", description: "list_files 返回的相对文件路径", minimum: nil, maximum: nil),
                "offset": .init(type: "integer", description: "从第几个字符开始，默认 0", minimum: 0, maximum: nil),
                "limit": .init(type: "integer", description: "最多读取的字符数，最大 50000", minimum: 1, maximum: 50_000)
            ], required: ["path"])
        )),
        ToolDefinition(function: .init(
            name: "search_files",
            description: "在已授权文件中做不区分大小写的文本搜索，返回文件、行号和片段。",
            parameters: .init(properties: [
                "query": .init(type: "string", description: "要查找的关键词或短语", minimum: nil, maximum: nil),
                "path": .init(type: "string", description: "可选的相对文件或目录，默认搜索全部", minimum: nil, maximum: nil)
            ], required: ["query"])
        )),
        ToolDefinition(function: .init(
            name: "write_file",
            description: "仅在用户要求生成文件时，将 UTF-8 文本写入 generated/。不能覆盖知识库原文或访问其他目录。",
            parameters: .init(properties: [
                "path": .init(type: "string", description: "必须以 generated/ 开头的相对路径", minimum: nil, maximum: nil),
                "content": .init(type: "string", description: "要写入的完整文本内容", minimum: nil, maximum: nil)
            ], required: ["path", "content"])
        ))
    ]

    static func completion(settings: AppSettings, messages: [Message], imagePNG: Data? = nil) async throws -> String {
        let data: Data
        if let imagePNG {
            data = try await perform(settings: settings, body: VisionRequest(
                model: await settings.model,
                messages: visionMessages(from: messages, imagePNG: imagePNG)
            ))
        } else {
            data = try await perform(settings: settings, body: Request(model: await settings.model, messages: messages))
        }
        guard let answer = try JSONDecoder().decode(Response.self, from: data).choices.first?.message.content else {
            throw AIError.emptyResponse
        }
        return answer
    }

    static func reactCompletion(settings: AppSettings, messages: [Message], workspace: KnowledgeFileTools,
                                imagePNG: Data? = nil, maxSteps: Int = 8) async throws -> ReActResult {
        var transcript = messages.map { ReActMessage(role: $0.role, content: $0.content) }
        let model = await settings.model
        for step in 0..<max(1, maxSteps) {
            let data: Data
            if step == 0, let imagePNG {
                data = try await perform(settings: settings, body: ReActVisionRequest(
                    model: model,
                    messages: reactVisionMessages(from: transcript, imagePNG: imagePNG),
                    tools: toolDefinitions
                ))
            } else {
                data = try await perform(settings: settings,
                                         body: ReActRequest(model: model, messages: transcript, tools: toolDefinitions))
            }
            guard let assistant = try JSONDecoder().decode(ReActResponse.self, from: data).choices.first?.message else {
                throw AIError.emptyResponse
            }
            transcript.append(assistant)
            if let calls = assistant.toolCalls, !calls.isEmpty {
                for call in calls {
                    let output = workspace.execute(name: call.function.name, arguments: call.function.arguments)
                    transcript.append(ReActMessage(role: "tool", content: output, toolCallID: call.id))
                }
                continue
            }
            if let answer = assistant.content?.trimmingCharacters(in: .whitespacesAndNewlines), !answer.isEmpty {
                return ReActResult(answer: answer, generatedFiles: workspace.generatedFiles.sorted())
            }
            throw AIError.emptyResponse
        }
        throw AIError.maxToolSteps
    }

    static func visionMessages(from messages: [Message], imagePNG: Data) -> [VisionMessage] {
        guard let target = messages.lastIndex(where: { $0.role == "user" }) else {
            return messages.map { VisionMessage(role: $0.role, content: .text($0.content)) }
        }
        return messages.enumerated().map { index, message in
            VisionMessage(role: message.role,
                          content: index == target ? .image(text: message.content, data: imagePNG) : .text(message.content))
        }
    }

    private static func reactVisionMessages(from messages: [ReActMessage], imagePNG: Data) -> [ReActVisionMessage] {
        let target = messages.lastIndex(where: { $0.role == "user" })
        return messages.enumerated().map { index, message in
            let content = message.content.map { value in
                index == target ? VisionMessage.Content.image(text: value, data: imagePNG) : .text(value)
            }
            return ReActVisionMessage(role: message.role, content: content,
                                      reasoningContent: message.reasoningContent, toolCalls: message.toolCalls,
                                      toolCallID: message.toolCallID)
        }
    }

    private static func perform<Body: Encodable>(settings: AppSettings, body: Body) async throws -> Data {
        let key = await settings.apiKey
        guard !key.isEmpty else { throw AIError.missingKey }
        let base = await settings.baseURL.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let endpoint = base.hasSuffix("/chat/completions") ? base : base + "/chat/completions"
        guard let url = URL(string: endpoint) else { throw AIError.invalidURL }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(key)", forHTTPHeaderField: "Authorization")
        request.httpBody = try JSONEncoder().encode(body)
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw AIError.server(String(data: data, encoding: .utf8) ?? "模型服务返回错误")
        }
        return data
    }
}

enum AIError: LocalizedError {
    case missingKey, invalidURL, emptyResponse, maxToolSteps, server(String)
    var errorDescription: String? {
        switch self {
        case .missingKey: "请先在设置中填写 API Key"
        case .invalidURL: "模型地址无效"
        case .emptyResponse: "模型没有返回内容"
        case .maxToolSteps: "模型连续调用工具次数过多，已停止本次任务"
        case .server(let message): message
        }
    }
}

File: src-tauri/src/ai_client.rs
use serde::{Deserialize, Serialize};
use crate::models::*;
use crate::errors::{AppError, AppResult};

#[derive(Serialize)]
struct ChatRequest {
    model: String,
    messages: Vec<ChatMsg>,
    temperature: f64,
}

#[derive(Serialize)]
struct VisionRequest {
    model: String,
    messages: Vec<VisionMsg>,
    temperature: f64,
}

#[derive(Serialize, Clone, Deserialize)]
pub struct ChatMsg {
    pub role: String,
    pub content: String,
}

#[derive(Serialize)]
struct VisionMsg {
    role: String,
    content: serde_json::Value,
}

#[derive(Deserialize)]
struct ChatResponse {
    choices: Vec<ChatChoice>,
}

#[derive(Deserialize)]
struct ChatChoice {
    message: ChatMsg,
}

pub async fn completion(settings: &AppSettings, messages: &[ChatMsg], image_png: Option<&[u8]>) -> AppResult<String> {
    let client = reqwest::Client::new();
    let endpoint = build_endpoint(&settings.base_url);

    let response = if let Some(img) = image_png {
        let vision_msgs = build_vision_messages(messages, img);
        let body = VisionRequest {
            model: settings.model.clone(),
            messages: vision_msgs,
            temperature: 0.2,
        };
        client.post(&endpoint)
            .header("Authorization", format!("Bearer {}", settings.api_key))
            .json(&body)
            .send()
            .await?
    } else {
        let body = ChatRequest {
            model: settings.model.clone(),
            messages: messages.to_vec(),
            temperature: 0.2,
        };
        client.post(&endpoint)
            .header("Authorization", format!("Bearer {}", settings.api_key))
            .header("Content-Type", "application/json")
            .json(&body)
            .send()
            .await?
    };

    if !response.status().is_success() {
        let status = response.status();
        let body = response.text().await.unwrap_or_default();
        return Err(AppError::Ai(format!("HTTP {}: {}", status, body)));
    }

    let result: ChatResponse = response.json().await?;
    result.choices.first()
        .map(|c| c.message.content.clone())
        .ok_or_else(|| AppError::Ai("Empty response".into()))
}

fn build_endpoint(base_url: &str) -> String {
    let trimmed = base_url.trim_end_matches('/');
    if trimmed.ends_with("/chat/completions") {
        trimmed.to_string()
    } else {
        format!("{}/chat/completions", trimmed)
    }
}

fn build_vision_messages(messages: &[ChatMsg], image_png: &[u8]) -> Vec<VisionMsg> {
    let target_idx = messages.iter().rposition(|m| m.role == "user");
    messages.iter().enumerate().map(|(i, msg)| {
        if Some(i) == target_idx {
            let b64 = base64::Engine::encode(&base64::engine::general_purpose::STANDARD, image_png);
            let content = serde_json::json!([
                {"type": "text", "text": msg.content},
                {"type": "image_url", "image_url": {"url": format!("data:image/png;base64,{}", b64)}}
            ]);
            VisionMsg { role: msg.role.clone(), content }
        } else {
            VisionMsg { role: msg.role.clone(), content: serde_json::json!(msg.content) }
        }
    }).collect()
}

// ReAct mode — tool-use loop
pub async fn react_completion(
    settings: &AppSettings,
    messages: &[ChatMsg],
    workspace: &mut crate::file_tools::KnowledgeFileTools,
    image_png: Option<&[u8]>,
    max_steps: usize,
) -> AppResult<ReActResult> {
    let client = reqwest::Client::new();
    let endpoint = build_endpoint(&settings.base_url);
    let mut transcript: Vec<ReActMsg> = messages.iter()
        .map(|m| ReActMsg { role: m.role.clone(), content: Some(m.content.clone()), tool_calls: None })
        .collect();

    for _step in 0..max_steps.max(1) {
        let body = ReActRequest {
            model: settings.model.clone(),
            messages: transcript.clone(),
            tools: tool_definitions(),
            temperature: 0.2,
        };

        let response = client.post(&endpoint)
            .header("Authorization", format!("Bearer {}", settings.api_key))
            .header("Content-Type", "application/json")
            .json(&body)
            .send()
            .await?;

        if !response.status().is_success() {
            return Err(AppError::Ai(format!("ReAct HTTP error: {}", response.status())));
        }

        let result: ReActResponse = response.json().await?;
        let assistant = result.choices.first()
            .map(|c| c.message.clone())
            .ok_or_else(|| AppError::Ai("Empty ReAct response".into()))?;

        transcript.push(assistant.clone());

        if let Some(ref calls) = assistant.tool_calls {
            if !calls.is_empty() {
                for call in calls {
                    let output = workspace.execute(&call.function.name, &call.function.arguments);
                    transcript.push(ReActMsg {
                        role: "tool".into(),
                        content: Some(output),
                        tool_calls: None,
                    });
                }
                continue;
            }
        }

        if let Some(ref answer) = assistant.content {
            let trimmed = answer.trim();
            if !trimmed.is_empty() {
                return Ok(ReActResult {
                    answer: trimmed.to_string(),
                    generated_files: workspace.generated_files().clone(),
                });
            }
        }
        return Err(AppError::Ai("Empty response".into()));
    }
    Err(AppError::Ai("Max tool steps exceeded".into()))
}

#[derive(Serialize, Clone, Deserialize)]
struct ReActMsg {
    role: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    content: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    tool_calls: Option<Vec<ToolCall>>,
}

#[derive(Serialize, Clone, Deserialize)]
struct ToolCall {
    id: String,
    #[serde(rename = "type")]
    call_type: String,
    function: ToolFunction,
}

#[derive(Serialize, Clone, Deserialize)]
struct ToolFunction {
    name: String,
    arguments: String,
}

#[derive(Serialize)]
struct ReActRequest {
    model: String,
    messages: Vec<ReActMsg>,
    tools: Vec<ToolDef>,
    temperature: f64,
}

#[derive(Serialize, Clone)]
struct ToolDef {
    #[serde(rename = "type")]
    def_type: String,
    function: ToolFnDef,
}

#[derive(Serialize, Clone)]
struct ToolFnDef {
    name: String,
    description: String,
    parameters: ToolParams,
}

#[derive(Serialize, Clone)]
struct ToolParams {
    #[serde(rename = "type")]
    param_type: String,
    properties: serde_json::Value,
    #[serde(skip_serializing_if = "Option::is_none")]
    required: Option<Vec<String>>,
}

#[derive(Deserialize)]
struct ReActResponse {
    choices: Vec<ReActChoice>,
}

#[derive(Deserialize)]
struct ReActChoice {
    message: ReActMsg,
}

#[derive(Debug, Clone)]
pub struct ReActResult {
    pub answer: String,
    pub generated_files: Vec<String>,
}

fn tool_definitions() -> Vec<ToolDef> {
    vec![
        ToolDef {
            def_type: "function".into(),
            function: ToolFnDef {
                name: "list_files".into(),
                description: "列出已授权的知识库文件".into(),
                parameters: ToolParams {
                    param_type: "object".into(),
                    properties: serde_json::json!({
                        "path": {
                            "type": "string",
                            "description": "可选的相对目录"
                        }
                    }),
                    required: None,
                },
            },
        },
        ToolDef {
            def_type: "function".into(),
            function: ToolFnDef {
                name: "read_file".into(),
                description: "按字符区间读取文件".into(),
                parameters: ToolParams {
                    param_type: "object".into(),
                    properties: serde_json::json!({
                        "path": {"type": "string", "description": "文件路径"},
                        "offset": {"type": "integer", "description": "起始字符", "minimum": 0},
                        "limit": {"type": "integer", "description": "最大字符数", "minimum": 1, "maximum": 50000}
                    }),
                    required: Some(vec!["path".into()]),
                },
            },
        },
        ToolDef {
            def_type: "function".into(),
            function: ToolFnDef {
                name: "search_files".into(),
                description: "在文件中搜索关键词".into(),
                parameters: ToolParams {
                    param_type: "object".into(),
                    properties: serde_json::json!({
                        "query": {"type": "string", "description": "搜索关键词"},
                        "path": {"type": "string", "description": "可选的文件或目录"}
                    }),
                    required: Some(vec!["query".into()]),
                },
            },
        },
        ToolDef {
            def_type: "function".into(),
            function: ToolFnDef {
                name: "write_file".into(),
                description: "写入文件到 generated/".into(),
                parameters: ToolParams {
                    param_type: "object".into(),
                    properties: serde_json::json!({
                        "path": {"type": "string", "description": "generated/ 下的路径"},
                        "content": {"type": "string", "description": "文件内容"}
                    }),
                    required: Some(vec!["path".into(), "content".into()]),
                },
            },
        },
    ]
}
